src/codegen/machine_ir.h

```diff
@@ -0,0 +1,180 @@
+  MIRRegister entry;
+  // Index within `instructions` of MIRFunction pointing to last
+  // instruction of basic block.
+  MIRRegister exit;
+
+  MIRFunction *function;
+
+  IRBlock *origin;
+} MIRBlock;
+
+typedef struct MIRFunction {
+  string name;
+
+  /// Flat list of all instructions in function.
+  MIRInstructionVector instructions;
+  MIRBlockVector blocks;
+
+  IRFunction *origin;
+} MIRFunction;
+
+/// dwisott
+MIRInstruction *mir_makenew(uint32_t opcode);
+/// Copy the entire instruction
+MIRInstruction *mir_makecopy(MIRInstruction *original);
+
+/// Clear the given instructions operands.
+void mir_op_clear(MIRInstruction *);
+
+/// Caller is responsible for calling vector_delete() on returned vector.
+MIRFunctionVector mir_from_ir(CodegenContext *context);
+
+/// If opcode is within common opcode range (less than or equal to
+/// MIR_COUNT), return a pointer to a NULL-terminated string containing
+/// a human-readable name of the opcode.
+const char *mir_common_opcode_mnemonic(uint32_t opcode);
+
+void print_mir_operand(MIROperand *m_op);
+void print_mir_instruction(MIRInstruction *m_inst);
+
+MIROperand mir_op_function(IRFunction *f);
+MIROperand mir_op_block(IRBlock *block);
+MIROperand mir_op_reference(MIRInstruction *inst);
+MIROperand mir_op_reference_ir(IRInstruction *inst);
+MIROperand mir_op_immediate(int64_t imm);
+MIROperand mir_op_name(const char *name);
+MIROperand mir_op_register(RegisterDescriptor reg, uint16_t size);
+
+void mir_add_op(MIRInstruction *inst, MIROperand op);
+/// Return a pointer to operand at index within instruction.
+MIROperand *mir_get_op(MIRInstruction *inst, size_t index);
+
+void mir_push(MIRFunction *mir, MIRInstruction *mi);
+MIRInstruction *mir_find_by_vreg(MIRFunction *mir, size_t reg);
+
+MIRFunction *mir_function(IRFunction *ir_f);
+MIRBlock *mir_block(MIRFunction *function, IRBlock *ir_bb);
+
+// MIRBlock *block, MIRInstruction *(it_name) <- name of iterator available in loop
+#define MIR_FOREACH_INST_IN_BLOCK(block, it_name)                       \
+  const size_t block##idx_start = (size_t)(block)->entry - (size_t)MIR_ARCH_START; \
+  const size_t block##idx_end = (size_t)(block)->exit - (size_t)MIR_ARCH_START; \
+  DBGASSERT(block##idx_start < block->function->instructions.size, "MIRBlock entry (%Z) is not less than size of machine instruction vector (%Z)", block##idx_start, block->function->instructions.size); \
+  DBGASSERT(block##idx_end < block->function->instructions.size, "MIRBlock exit (%Z) larger than size of machine instruction vector (%Z) of function %S", block##idx_end, block->function->instructions.size, block->function->name); \
+  DBGASSERT(block##idx_start <= block##idx_end, "MIRBlock exit is before it's entry: %Z < %Z", block##idx_end, block##idx_start); \
+  MIRInstruction *it_name = block->function->instructions.data[block##idx_start]; \
+  for (size_t it = block##idx_start; it <= block##idx_end; ++it, it_name = block->function->instructions.data[it])
+
```

Comment:
This is *probably* fine, but I’d generally always use `CAT` rather than `##` directly, just in case `block` or `it_name` ends up being a macro.

Reply:
Also, the possessive form of ‘it’ is ‘its’, and not ‘it’s’ ;Þ

Reply:
That’s strange. It should definitely work...